core/src/main/java/org/apache/iceberg/io/CDCWriter.java

```diff
@@ -0,0 +1,135 @@
+  private final Map<StructLike, PartitionAwarePathOffset> insertedRows;
+  private final PositionDelete<T> positionDelete;
+  private final Function<T, StructLike> toStructLike;
+
+  public CDCWriter(FanoutDataWriter<T> dataWriter,
+                   PartitionAwareFileWriter<T, DeleteWriteResult> equalityDeleteWriter,
+                   PartitionAwareFileWriter<PositionDelete<T>, DeleteWriteResult> positionDeleteWriter,
+                   Schema schema, Schema deleteSchema, Function<T, StructLike> toStructLike) {
+    this.dataWriter = dataWriter;
+    this.equalityDeleteWriter = equalityDeleteWriter;
+    this.positionDeleteWriter = positionDeleteWriter;
+    this.positionDelete = new PositionDelete<>();
+    this.keyProjection = StructProjection.create(schema, deleteSchema);
+    this.insertedRows = StructLikeMap.create(deleteSchema.asStruct());
+    this.toStructLike = toStructLike;
+  }
+
+  @Override
+  public void insert(T row, PartitionSpec spec, StructLike partition) throws IOException {
+    CharSequence currentPath = dataWriter.currentPath(spec, partition);
+    long currentPosition = dataWriter.currentPosition(spec, partition);
+    PartitionAwarePathOffset offset = new PartitionAwarePathOffset(spec, partition, currentPath, currentPosition);
+
+    StructLike copiedKey = StructCopy.copy(keyProjection.wrap(toStructLike.apply(row)));
+
+    PartitionAwarePathOffset previous = insertedRows.put(copiedKey, offset);
+    if (previous != null) {
+      // TODO: attach the previous row if has a position delete row schema
+      positionDelete.set(previous.path(), previous.rowOffset(), null);
+      positionDeleteWriter.write(positionDelete, spec, partition);
+    }
+
+    dataWriter.write(row, spec, partition);
+  }
+
+  @Override
+  public void delete(T row, PartitionSpec spec, StructLike partition) throws IOException {
+    StructLike key = keyProjection.wrap(toStructLike.apply(row));
+    PartitionAwarePathOffset previous = insertedRows.remove(key);
+    if (previous != null) {
+      // TODO: attach the previous row if has a position delete row schema
+      positionDelete.set(previous.path(), previous.rowOffset(), null);
+      positionDeleteWriter.write(positionDelete, previous.spec, previous.partition);
+    }
+
+    equalityDeleteWriter.write(row, spec, partition);
+  }
+
+  @Override
+  public void delete(CharSequence path, long pos, T row, PartitionSpec spec, StructLike partition) throws IOException {
+    throw new IllegalArgumentException(this.getClass().getName() + " does not implement explicit position delete");
+  }
+
+  @Override
+  protected void closeWriters() throws IOException {
+    if (dataWriter != null) {
+      closeDataWriter(dataWriter);
+    }
+
+    if (equalityDeleteWriter != null) {
+      closeDeleteWriter(equalityDeleteWriter);
+    }
+
+    if (positionDeleteWriter != null) {
+      closeDeleteWriter(positionDeleteWriter);
+    }
+  }
+
+  private static class PartitionAwarePathOffset {
```

Comment:
Could this extend `PositionDelete` so it can be passed directly into the delete writer?